### PR TITLE
Adds support for XDG base directories on Linux

### PR DESCRIPTION
**Pull request by 3ster**
related issue: #18 
Includes $XDG_PICTURES_DIR for screenshots, $XDG_CONFIG_DIR for
configurations, and $XDG_DATA_DIR for game data.
Game data is loaded from $XDG_DATA_DIR/legacyclonk first,
/usr/share/legacyclonk second.
In my (admittedly limited) testing this worked without issues, including being able to load different scenarios.

### CODE REVIEW

**Comment by Somebodyisnobody** [comment]
@maxmitti similar to my proposal for Windows to use AppData for content.

File: src/C4Config.cpp [hunk 1]
```diff
@@ -16,6 +16,7 @@
 
 /* Game configuration as stored in registry */
 
+
```
Comment (Fulgen301) [inline src/C4Config.cpp:19]: Why is that needed?

File: src/C4Config.cpp [hunk 2]
```diff
@@ -75,7 +76,7 @@ void C4ConfigGeneral::CompileFunc(StdCompiler *pComp)
 #ifdef _WIN32
 	pComp->Value(mkNamingAdapt(s(UserPath), "UserPath", "%APPDATA%\\LegacyClonk",                        false, true));
 #elif defined(__linux__)
-	pComp->Value(mkNamingAdapt(s(UserPath), "UserPath", "$HOME/.legacyclonk",                            false, true));
+	pComp->Value(mkNamingAdapt(s(UserPath), "UserPath", "$XDG_CONFIG_HOME/.legacyclonk",                            false, true));
```
Comment (Fulgen301) [inline src/C4Config.cpp:79]: The indentation of the other arguments is messed up here. Whether that indentation's needed might be open up for debate, but we shouldn't break it at only one place.

File: src/C4Config.cpp [hunk 3]
```diff
@@ -378,6 +379,7 @@ bool C4Config::Load(bool forceWorkingDirectory, const char *szConfigFile)
 		{
 			// Nonwindows or explicit config file: Determine filename to load config from
 			StdStrBuf filename;
+			StdStrBuf config_path;
```
Comment (Fulgen301) [inline src/C4Config.cpp:382]: This does not follow the style guide (`configPath` instead of `config_path`).

File: src/C4Config.cpp [hunk 4]
```diff
@@ -387,14 +389,22 @@ bool C4Config::Load(bool forceWorkingDirectory, const char *szConfigFile)
 			}
 			else
 			{
+#ifdef __APPLE__
```
Comment (Fulgen301) [inline src/C4Config.cpp:392]: This isn't apple-specific stuff, just code which doesn't use XDG. BSD would use the wrong code path as well.

File: src/C4Config.cpp [hunk 5]
```diff
@@ -387,14 +389,22 @@ bool C4Config::Load(bool forceWorkingDirectory, const char *szConfigFile)
...
 				// Config filename from home
 				StdStrBuf home(getenv("HOME"), false);
 				if (home) { home += "/"; }
 				filename.Copy(home);
-#ifdef __APPLE__
 				filename += "Library/Preferences/legacyclonk.config";
 #else
-				filename += ".legacyclonk/config";
+				// get XDG config folder instead of polluting ~
+				StdStrBuf xdg_config_home(getenv("XDG_CONFIG_HOME"));
```
Comment (Fulgen301) [inline src/C4Config.cpp:400]: The style guide mandates brace initialization.

File: src/C4Config.cpp [hunk 6]
```diff
@@ -497,12 +504,20 @@ bool C4Config::Save()
 			}
 			else
 			{
+#ifndef __linux__
 				filename.Copy(getenv("HOME"));
+#else
+				filename.Copy(Config.General.XDG_CONFIG_PATH);
```
Comment (Fulgen301) [inline src/C4Config.cpp:510]: Why XDG_CONFIG_PATH? We might need a special ConfigPath for other platforms as well, so it shouldn't have a platform dependent name.

File: src/C4Config.cpp [hunk 7]
```diff
@@ -497,12 +504,20 @@ bool C4Config::Save()
 			}
 			else
 			{
+#ifndef __linux__
```
Comment (Fulgen301) [inline src/C4Config.cpp:507]: Same as above. XDG isn't Linux-only.

File: src/C4Config.cpp [hunk 8]
```diff
@@ -497,12 +504,20 @@ bool C4Config::Save()
...
 				filename.Copy(getenv("HOME"));
+#else
+				filename.Copy(Config.General.XDG_CONFIG_PATH);
+#endif
 				if (filename) { filename += "/"; }
 #ifdef __APPLE__
 				filename += "Library/Preferences/legacyclonk.config";
 #else
+#ifndef __linux__
```
Comment (Fulgen301) [inline src/C4Config.cpp:516]: Same as above.

File: src/C4Config.cpp [hunk 9]
```diff
@@ -547,6 +562,34 @@ void C4ConfigGeneral::DeterminePaths(bool forceWorkingDirectory)
...
+	}
+	SCopy(xdg_config_path.getData(), XDG_CONFIG_PATH);
+    SCopy(xdg_config_path.getData(), UserPath);
+    SCopy(xdg_config_path.getData(), LogPath); // I'm not sure if this even gets used, but setting it makes the config look nicer :-)
+    StdStrBuf xdg_picture_path(getenv("XDG_PICTURES_DIR"));
+    if (!xdg_picture_path.getSize()){
+        xdg_picture_path = getenv("HOME");
+        xdg_picture_path += "/Pictures";
+    }
+    struct stat st;
```
Comment (Fulgen301) [inline src/C4Config.cpp:578]: This is C++, so `struct Type var` isn't needed, `Type var` is enough.

File: src/C4Config.cpp [hunk 10]
```diff
@@ -547,6 +562,34 @@ void C4ConfigGeneral::DeterminePaths(bool forceWorkingDirectory)
...
 	else
 		SCopy("/tmp/", TempPath);
+	StdStrBuf xdg_config_path(getenv("XDG_CONFIG_HOME"));
+	if (!xdg_config_path.getSize()){
+	    xdg_config_path = getenv("HOME");
+	    xdg_config_path += "/.config/legacyclonk";
+	}
+	SCopy(xdg_config_path.getData(), XDG_CONFIG_PATH);
+    SCopy(xdg_config_path.getData(), UserPath);
+    SCopy(xdg_config_path.getData(), LogPath); // I'm not sure if this even gets used, but setting it makes the config look nicer :-)
```
Comment (Fulgen301) [inline src/C4Config.cpp:572]: Doing so breaks third-party programs assuming the log path to be equal to the executable path. We might not care about it though.

File: src/C4Config.cpp [hunk 11]
```diff
@@ -547,6 +562,34 @@ void C4ConfigGeneral::DeterminePaths(bool forceWorkingDirectory)
...
+        xdg_picture_path += "/Pictures";
+    }
+    struct stat st;
+    if (!((stat(xdg_picture_path.getData(), &st) == 0) && S_ISDIR(st.st_mode))) {
+        // fall back to saving screenshots in config folder
+        xdg_picture_path = xdg_config_path;
+    }
+    else {
+        xdg_picture_path += "/legacyclonk";
+        if (!((stat(xdg_picture_path.getData(), &st) == 0) && S_ISDIR(st.st_mode))) {
```
Comment (Fulgen301) [inline src/C4Config.cpp:585]: StdFile.cpp contains `DirectoryExists`, no need to reimplement the wheel.

File: src/C4Config.cpp [hunk 12]
```diff
@@ -547,6 +562,34 @@ void C4ConfigGeneral::DeterminePaths(bool forceWorkingDirectory)
...
+    SCopy(xdg_config_path.getData(), LogPath); // I'm not sure if this even gets used, but setting it makes the config look nicer :-)
+    StdStrBuf xdg_picture_path(getenv("XDG_PICTURES_DIR"));
+    if (!xdg_picture_path.getSize()){
+        xdg_picture_path = getenv("HOME");
+        xdg_picture_path += "/Pictures";
+    }
+    struct stat st;
+    if (!((stat(xdg_picture_path.getData(), &st) == 0) && S_ISDIR(st.st_mode))) {
+        // fall back to saving screenshots in config folder
+        xdg_picture_path = xdg_config_path;
```
Comment (Fulgen301) [inline src/C4Config.cpp:581]: That doesn't look right - Screenshots in your config folder? Apart from that being the completely wrong place, there's no subdirectory either.

File: src/C4Config.cpp [hunk 13]
```diff
@@ -578,9 +625,21 @@ void C4ConfigGeneral::DeterminePaths(bool forceWorkingDirectory)
 
 char AtPathFilename[_MAX_PATH + 1];
 
-const char *C4Config::AtExePath(const char *szFilename)
+const char *C4Config::AtExePath(const char *szFilename, bool forceExeFolder)
```
Comment (Fulgen301) [inline src/C4Config.cpp:628]: Umm, no. That function explicitely returns a path to a file inside the executable directory. It must *never* return something else, otherwise we could throw it out of the window. This isn't the right place to implement the required functionality.

File: src/C4Config.cpp [hunk 14]
```diff
@@ -663,10 +722,9 @@ void C4ConfigControls::ResetKeys()
 	StdCompilerNull Comp; Comp.Compile(mkParAdapt(*this, true));
 }
 
-const char *C4Config::AtExeRelativePath(const char *szFilename)
-{
-	// Specified file is located in ExePath: return relative path
-	return GetRelativePathS(szFilename, General.ExePath);
+const char *C4Config::AtExeRelativePath(const char *szFilename) {
```
Comment (Fulgen301) [inline src/C4Config.cpp:725]: The style guide mandates a specific brace style, and this change doesn't even do anything else.

File: src/C4Config.h [hunk 15]
```diff
@@ -67,6 +67,9 @@ class C4ConfigGeneral
 	int32_t ConfigResetSafety; // safety value: If this value is screwed, the config got currupted and must be reset
 	// Determined at run-time
 	char ExePath[CFG_MaxString + 1];
+#ifdef __linux__
+	char XDG_CONFIG_PATH[CFG_MaxString + 1];
```
Comment (Fulgen301) [inline src/C4Config.h:71]: For new code, please use `std::string`. Also, platform dependency.

File: src/C4Config.h [hunk 16]
```diff
@@ -267,7 +270,7 @@ class C4Config
 	bool Save();
 	bool Load(bool forceWorkingDirectory = true, const char *szConfigFile = nullptr);
 	bool Init();
-	const char *AtExePath(const char *szFilename);
+	const char *AtExePath(const char *szFilename, bool forceExeFolder=false);
```
Comment (Fulgen301) [inline src/C4Config.h:273]: See other comment above. Also, the style guide mandates spaces next to the operator.

File: src/C4Group.cpp [hunk 17]
```diff
@@ -670,6 +670,35 @@ bool C4Group::Open(const char *szGroupName, bool fCreate)
...
+	    if (FileExists(filename.getData()))
+        {
+	        Init();
+	        return OpenReal(filename.getData());
+        }
+
+	}
+#endif
+
+
```
Comment (Fulgen301) [inline src/C4Group.cpp:701]: Sorry, no. C4Group isn't even remotely the right place for this. C4Group should handle file IO, as it already does - it *must not* deal with which operating system or configuration wants what file to be where. Just no. This is also a way to make C4Group access files regardless of what path is actually required. The more I think about it, the worse it gets. Please don't do that.

Also, you hardcoded legacyclonk - the config does this too (which is equally bad, but not your fault).

File: src/C4Log.cpp [hunk 18]
```diff
@@ -37,15 +37,28 @@ StdStrBuf sFatalError;
...
 #else
 	while (!(C4LogFile = fopen(sLogFileName.getData(), "wb")))
 #endif
 	{
 		// try different name
+#ifndef __linux__
 		sLogFileName.Format(C4CFN_LogEx, iLog++);
+#else
+		sLogFileName = Config.General.XDG_CONFIG_PATH;
+		sLogFileName.Append("/");
```
Comment (Fulgen301) [inline src/C4Log.cpp:59]: Why do this twice?

File: src/C4StartupMainDlg.cpp [hunk 19]
```diff
@@ -283,7 +283,13 @@ void C4StartupMainDlg::OnShown()
 	// first thing that's needed is a new player, if there's none - independent of first start
 	bool fHasPlayer = false;
 	StdStrBuf sSearchPath; const char *szFn;
+#ifndef __linux__
 	sSearchPath.Format("%s%s", (const char *)Config.General.ExePath, (const char *)Config.General.PlayerPath);
+#else
+	// search in user config directory instead.
+	sSearchPath.Format("%s%s", (const char*)Config.General.XDG_CONFIG_PATH, (const char*)Config.General.PlayerPath);
```
Comment (Fulgen301) [inline src/C4StartupMainDlg.cpp:290]: Please use C++-casts for new code.

File: src/C4StartupPlrSelDlg.cpp [hunk 20]
```diff
@@ -694,7 +694,12 @@ void C4StartupPlrSelDlg::UpdatePlayerList()
 		// player mode: insert all players
 		const char *szFn;
 		StdStrBuf sSearchPath;
+#ifndef __linux__
 		sSearchPath.Format("%s%s", (const char *)Config.General.ExePath, (const char *)Config.General.PlayerPath);
+#else
+        sSearchPath.Format("%s%s", (const char *)Config.General.XDG_CONFIG_PATH, (const char *)Config.General.PlayerPath);
```
Comment (Fulgen301) [inline src/C4StartupPlrSelDlg.cpp:700]: Indentation error.

File: src/C4StartupScenSelDlg.cpp [hunk 21]
```diff
@@ -1400,6 +1400,21 @@ void C4StartupScenSelDlg::OnShown()
...
+	}
+    data_folder += "/";
+	pScenLoader->Load(data_folder);
+	UpdateList();
+	UpdateSelection();
+	data_folder = "/usr/share/legacyclonk/";
+	pScenLoader->Load(data_folder);
+	UpdateList();
+	UpdateSelection();
+#endif
```
Comment (Fulgen301) [inline src/C4StartupScenSelDlg.cpp:1417]: That's a lot of duplicate code for no good reason. I also don't like that scenarios should be in the **configuration** directory - what?

**Comment by 3ster** [comment]
Upon further testing it seems like loading scenarios from ExeDir, $XDG_DATA_DIR/legacyclonk, *and* /usr/share/legacyclonk currently does not work as the scenario list is cleared on every new Load.
As such, only scenarios from /usr/share/legacyclonk are loaded as of right now.

File: src/C4StartupScenSelDlg.cpp (hunk 21 above)
Comment (3ster) [reply src/C4StartupScenSelDlg.cpp:1417]: Using XDG_CONFIG_DATA here instead of XDG_DATA_HOME here is a brain issue on my side - the fallback code actually loads scenarios from the (proper) local data directory.

File: src/C4Log.cpp (hunk 18 above)
Comment (3ster) [reply src/C4Log.cpp:59]: Once for Clonk.log, once for Clonk[number].log. I'm sure there's a better way to do this.

Comment (Fulgen301) [reply src/C4Log.cpp:59]: Yes, but you can cache the result of the config path with the slash appended.

File: src/C4Config.cpp (hunk 4 above)
Comment (3ster) [reply src/C4Config.cpp:392]: I initially had different code here, looks like I messed up restoring it to its previous form here.

**Comment by 3ster** [comment]
Thank you for your in-depth code review.
While I blame most of the style and indentation issues on auto formatting and missing insight into the (apparently non-public?) style guide, I agree that there's more pressing issues here.
I'm currently still working on fixing these aforementioned issues, including the one mentioned in my comment above.

**Comment by maxmitti** [comment]
Before you work further on addressing the code review, I would like to make a suggestion.
Even though the old code has some platform specific code in C4Config (and other places), it would be much cleaner to have a separate class which only deals with mapping paths to the right platform dependent paths.
This would be especially useful as it would already solve changes for other platforms in the future.

Anyway, this is only relevant if the stuff is not backported from OC.

As for the style guide, as far as I know the one who said would create the style guide didn't publish it yet because it isn't done yet. Which is a shame, seeing this situation now.

That said, I appreciate your interest.
I was quite surprised to see someone interested who isn't part of the usual team.

**Comment by maxmitti** [comment]
Closed due to inactivity and in favor of #36.
If you are still interested, just comment here or in the issue.
